glom/core.py

```diff
@@ -1323,13 +1323,35 @@ def _t_child(parent, operation, arg):
     return t
 
 
+def _s_first_magic(scope, key, _t):
+    """
+    enable S.a to do S['a'] or S['a'].val as a special
```

Comment:
the .val comes as a bit of a surprise. Is it not possible to have it simply be the value? Are there times when the user would want the Var instance? Spec authors have access to the real scope if there's fancy stuff to do. I saw Var as a declaration, and lack of a Var up the scope chain to basically mean implicitly putting a Var() in the current scope (if only semantically, no need for the overhead).

Reply:
I think putting the `_VarVal` container in place is the easiest way to do the following:

1- quickly error if assigning to undeclared variable

2- quickly error if reading from unassigned variable

3- manage scope / lifetime of value correctly

Imagine the mechanics of getting this right:

`(Let(a=Var()), Let(a=1), Let(a=2))`

So, first you put `a=Var()` in the spec.

Then `Let(a=1)`, so you backtrack up and see the `Var()`.  And then... what?

If you replace the `Var()` with 1, now `Let(a=2)` will not see the `Var()` when it backtracks.

So you need to durably record the allocation somehow.

Reply:
As for doing fancy things and wanting to access the "raw" scope -- no problem!!

1- `S` is not the scope, the ACTUAL scope inside the spec is unaffected this is just some T/path magic

2- `S[]` will access everything raw -- `S.a` might unpack the values, but `S['a']` will not